enterprise/server/bes_artifacts/bes_artifacts.go

```diff
@@ -0,0 +1,159 @@
+package bes_artifacts
+
+import (
+	"context"
+	"fmt"
+	"io/fs"
+	"net/url"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/build_event_publisher"
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/cachetools"
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/digest"
+	"github.com/buildbuddy-io/buildbuddy/server/util/grpc_client"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"golang.org/x/sync/errgroup"
+
+	bespb "github.com/buildbuddy-io/buildbuddy/proto/build_event_stream"
+	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+	bspb "google.golang.org/genproto/googleapis/bytestream"
+)
+
+type Result struct {
+	Duration   time.Duration
+	Digest     *repb.Digest
+	NamedSetID string
+	Path       string
+	Name       string
+	Err        error
+}
+
+// Uploader can be used to asynchronously upload artifacts associated with a
+// build event stream.
+type Uploader struct {
+	ctx context.Context
+	eg  *errgroup.Group
+	// ByteStream client used to upload artifacts.
+	bsClient bspb.ByteStreamClient
+	// Publisher used to publish NamedSetOfFiles events when background uploads
+	// have completed.
+	bep *build_event_publisher.Publisher
+	// Bytestream URI prefix (like "bytestream://remote.buildbuddy.io:443")
+	bytestreamURIPrefix string
+	// Remote instance name to be used for uploaded artifacts.
+	instanceName string
+
+	resultsMu sync.Mutex
+	// Array containing all upload results.
+	results []*Result
+}
+
+func NewUploader(ctx context.Context, bep *build_event_publisher.Publisher, cacheTarget, instanceName string) (*Uploader, error) {
+	eg, ctx := errgroup.WithContext(ctx)
+
+	u, err := url.Parse(cacheTarget)
+	if err != nil {
+		return nil, status.InvalidArgumentErrorf("failed to parse cache target %q: %s", cacheTarget, err)
+	}
+	bytestreamURIPrefix := "bytestream://" + u.Host
+
+	conn, err := grpc_client.DialTarget(cacheTarget)
+	if err != nil {
+		return nil, err
+	}
+	go func() {
+		<-ctx.Done()
+		conn.Close()
+	}()
```

Comment:
wdyt about doing this in Wait()?

That way the lifetime of this object is more explicit and likely to be shorter. It also avoids a leak if someone calls it with a background context or something silly.

Reply:
Done